File: app/src/main/java/de/westnordost/streetcomplete/quests/ImageListQuestAnswerFragment.java
package de.westnordost.streetcomplete.quests;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

import javax.inject.Inject;

import de.westnordost.streetcomplete.Injector;
import de.westnordost.streetcomplete.R;
import de.westnordost.streetcomplete.view.ImageSelectAdapter;
import de.westnordost.streetcomplete.view.Item;

/**
 * Abstract class for quests with a list of images and one or several to select.
 */
public abstract class ImageListQuestAnswerFragment extends AbstractQuestFormAnswerFragment {

	public static final String OSM_VALUES = "osm_values";

    private static final String
			SELECTED_INDICES = "selected_indices",
			EXPANDED = "expanded";

	protected ImageSelectAdapter imageSelector;
    private Button showMoreButton;
	private RecyclerView valueList;

	private List<Item> allItems;

	@Inject LastPickedValuesStore favs;

	@Override public void onCreate(Bundle inState)
	{
		super.onCreate(inState);
		Injector.instance.getApplicationComponent().inject(this);
		allItems = Collections.unmodifiableList(Arrays.asList(getItems()));
	}

    @Override public View onCreateView(LayoutInflater inflater, ViewGroup container,
                                       Bundle savedInstanceState)
    {
        View view = super.onCreateView(inflater, container, savedInstanceState);

        View contentView = setContentView(R.layout.quest_generic_list);

		valueList = contentView.findViewById(R.id.listSelect);
        GridLayoutManager lm = new GridLayoutManager(getActivity(), getItemsPerRow());
        valueList.setLayoutManager(lm);
		valueList.setNestedScrollingEnabled(false);

		showMoreButton = contentView.findViewById(R.id.buttonShowMore);
		showMoreButton.setOnClickListener(v ->
		{
			imageSelector.setItems(getItems(-1));
			showMoreButton.setVisibility(View.GONE);
		});

		int selectableItems = getMaxSelectableItems();
		TextView selectHint = contentView.findViewById(R.id.selectHint);
		selectHint.setText(selectableItems == 1 ? R.string.quest_roofShape_select_one : R.string.quest_select_hint);

		imageSelector = new ImageSelectAdapter(selectableItems);
		imageSelector.addOnItemSelectionListener(new ImageSelectAdapter.OnItemSelectionListener()
		{
			@Override public void onIndexSelected(int index) { checkIsFormComplete(); }
			@Override public void onIndexDeselected(int index) { checkIsFormComplete(); }
		});

        return view;
    }

	@Override public void onViewCreated(View view, @Nullable Bundle savedInstanceState)
	{
		super.onViewCreated(view, savedInstanceState);

		int initiallyShow = getMaxNumberOfInitiallyShownItems();
		if(savedInstanceState != null)
		{
			if(savedInstanceState.getBoolean(EXPANDED)) initiallyShow = -1;
			showItems(initiallyShow);

			List<Integer> selectedIndices = savedInstanceState.getIntegerArrayList(SELECTED_INDICES);
			imageSelector.select(selectedIndices);
		}
		else
		{
			showItems(initiallyShow);
		}
		valueList.setAdapter(imageSelector);
	}

	protected int getItemsPerRow()
	{
		return 4;
	}
	/** return -1 for any number. Default: 1 */
    protected int getMaxSelectableItems()
	{
		return 1;
	}
	/** return -1 for showing all items at once. Default: -1 */
	protected int getMaxNumberOfInitiallyShownItems()
	{
		return -1;
	}
	protected abstract Item[] getItems();

	@Override protected void onClickOk()
	{
		applyAnswer();
	}

	protected void applyAnswer()
	{
		Bundle answer = new Bundle();

		ArrayList<String> osmValues = new ArrayList<>();
		for(Item item : imageSelector.getSelectedItems())
		{
			osmValues.add(item.value);
		}
		if(!osmValues.isEmpty())
		{
			answer.putStringArrayList(OSM_VALUES, osmValues);
		}
		favs.addLastPicked(getClass().getSimpleName(), osmValues);
		applyAnswer(answer);
	}

    @Override public void onSaveInstanceState(Bundle outState)
    {
        super.onSaveInstanceState(outState);
		outState.putIntegerArrayList(SELECTED_INDICES, imageSelector.getSelectedIndices());
		outState.putBoolean(EXPANDED, showMoreButton.getVisibility() == View.GONE);
    }

    @Override public boolean isFormComplete()
    {
        return !imageSelector.getSelectedIndices().isEmpty();
    }

	private void showItems(int initiallyShow)
	{
		if(initiallyShow == -1 || initiallyShow >= allItems.size())
		{
			showMoreButton.setVisibility(View.GONE);
		}

		imageSelector.setItems(getItems(initiallyShow));
	}

	private List<Item> getItems(int showOnly)
	{
		LinkedList<Item> items;
		if(showOnly == -1 || showOnly >= allItems.size())
		{
			items = new LinkedList<>(allItems);
		}
		else
		{
			items = new LinkedList<>(allItems.subList(0, showOnly));
		}

		if(allItems.size() > getItemsPerRow())
		{
			favs.moveLastPickedToFront(getClass().getSimpleName(), items, allItems);
		}
		return items;
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/quests/ImageListQuestAnswerFragment.kt
package de.westnordost.streetcomplete.quests

import android.os.Bundle
import androidx.recyclerview.widget.GridLayoutManager
import android.view.View
import androidx.core.os.bundleOf

import java.util.ArrayList
import java.util.LinkedList

import javax.inject.Inject

import de.westnordost.streetcomplete.Injector
import de.westnordost.streetcomplete.R
import de.westnordost.streetcomplete.view.GroupedItem
import de.westnordost.streetcomplete.view.ImageSelectAdapter
import kotlinx.android.synthetic.main.quest_generic_list.*

/**
 * Abstract class for quests with a list of images and one or several to select.
 */
abstract class ImageListQuestAnswerFragment : AbstractQuestFormAnswerFragment() {

    override val contentLayoutResId = R.layout.quest_generic_list

    protected lateinit var imageSelector: ImageSelectAdapter

    @Inject internal lateinit var favs: LastPickedValuesStore

    protected open val itemsPerRow = 4
    /** return -1 for any number. Default: 1  */
    protected open val maxSelectableItems = 1
    /** return -1 for showing all items at once. Default: -1  */
    protected open val maxNumberOfInitiallyShownItems = -1

    protected abstract val items: List<GroupedItem>

    init {
        Injector.instance.applicationComponent.inject(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        imageSelector = ImageSelectAdapter(maxSelectableItems)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        list.layoutManager = GridLayoutManager(activity, itemsPerRow)
        list.isNestedScrollingEnabled = false

        selectHintLabel.setText(if (maxSelectableItems == 1) R.string.quest_roofShape_select_one else R.string.quest_select_hint)

        imageSelector.listeners.add(object : ImageSelectAdapter.OnItemSelectionListener {
            override fun onIndexSelected(index: Int) {
                checkIsFormComplete()
            }

            override fun onIndexDeselected(index: Int) {
                checkIsFormComplete()
            }
        })

        showMoreButton.setOnClickListener {
            imageSelector.items = items.favouritesMovedToFront()
            showMoreButton.visibility = View.GONE
        }

        var initiallyShow = maxNumberOfInitiallyShownItems
        if (savedInstanceState != null) {
            if (savedInstanceState.getBoolean(EXPANDED)) initiallyShow = -1
            showItems(initiallyShow)

            val selectedIndices = savedInstanceState.getIntegerArrayList(SELECTED_INDICES)!!
            imageSelector.select(selectedIndices)
        } else {
            showItems(initiallyShow)
        }
        list.adapter = imageSelector
    }

    override fun onClickOk() {
        applyAnswer()
    }

    protected fun applyAnswer() {
        val osmValues = imageSelector.selectedItems.map { it.value!! }
        if (osmValues.isNotEmpty()) {
            favs.add(javaClass.simpleName, osmValues)
            applyAnswer(bundleOf(OSM_VALUES to ArrayList(osmValues)))
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putIntegerArrayList(SELECTED_INDICES, ArrayList(imageSelector.selectedIndices))
        outState.putBoolean(EXPANDED, showMoreButton.visibility == View.GONE)
    }

    override fun isFormComplete() = imageSelector.selectedIndices.isNotEmpty()

    private fun showItems(initiallyShow: Int) {
        val allItems = items
        val showAll = initiallyShow == -1 || initiallyShow >= allItems.size

        showMoreButton.visibility = if(showAll) View.GONE else View.VISIBLE
        val sortedItems = allItems.favouritesMovedToFront()
        imageSelector.items = if(showAll) sortedItems else sortedItems.subList(0, initiallyShow)
    }

    private fun List<GroupedItem>.favouritesMovedToFront(): List<GroupedItem> {
        val result: LinkedList<GroupedItem> = LinkedList(this)

        if (result.size > itemsPerRow) {
            favs.moveLastPickedToFront(javaClass.simpleName, result, this)
        }
        return result
    }

    companion object {

        const val OSM_VALUES = "osm_values"

        private const val SELECTED_INDICES = "selected_indices"
        private const val EXPANDED = "expanded"
    }
}
